### PR TITLE
user profile page

### PR DESCRIPTION
**Pull request by SetterVet**

### CODE REVIEW

**Comment by SetterVet** [comment]
Make json for user profile

**Comment by SetterVet** [comment]
image

File: app/models/user.rb [hunk 1]
```diff
@@ -47,4 +47,20 @@ def self.from_omniauth(access_token)
     end
     user
   end
+
+  def self.event_as_json(user)
```
Comment (VasylYaremkiv) [inline app/models/user.rb]: Who uses this method?

Comment (VasylYaremkiv) [reply app/models/user.rb]: Move it for event models and pass `events`

**Comment by SetterVet** [comment]
image

**Comment by SetterVet** [comment]
image

**Comment by SetterVet** [comment]
image

**Comment by SetterVet** [comment]
Ready to squash. Please check @VasylYaremkiv @sonishc @afurm @rubycop

File: app/javascript/components/users/UserEvents.vue [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            </div>
+            <div class="col-2">
+                <p class="event-finish-date">{{ event["finish_date"] }}</p>
+            </div>
+            <div class="col-10">
+                <div class="row event-visible-item">
+                    <div class="col-8" @click="changeActiveEvent(event.id)">
+                        <span>{{ event.description }}</span>
+                    </div>
+                    <div class="col-2" @click="alert('h')">
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/UserEvents.vue]: Remove `alert`

File: app/javascript/components/users/UserEvents.vue [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                if (this.getFullYear(event) === this.topYear ){
+                    return true
+                }
+                else {
+                    this.topYear = this.getFullYear(event);
+                    return false
+                }
+            },
+            changeActiveEvent (eventID){
+                let hidden_event_info = document.getElementById('event'+eventID);
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/UserEvents.vue]: Use camelCase for variables: https://github.com/airbnb/javascript#naming--camelCase

Comment (VasylYaremkiv) [reply app/javascript/components/users/UserEvents.vue]: Add missed spaces

File: app/javascript/components/users/UserEvents.vue [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            topEqualCurrent(event) {
+                if (this.getFullYear(event) === this.topYear ){
+                    return true
+                }
+                else {
+                    this.topYear = this.getFullYear(event);
+                    return false
+                }
+            },
+            changeActiveEvent (eventID){
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/UserEvents.vue]: `eventId`

File: app/javascript/components/users/UserEvents.vue [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+</template>
+
+<script>
+    export default {
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/UserEvents.vue]: In all components uses 2 spaces, but you used 4. Fix it.

File: app/javascript/src/styles/pages/_user.scss [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  cursor: pointer;
+}
+
+.event-status {
+  border-radius: 5px;
+  padding-left: 10px;
+  background: #0a568c;
+  margin-top: 10px;
+}
+.scheduled {
```
Comment (VasylYaremkiv) [inline app/javascript/src/styles/pages/_user.scss]: Add empty line between css rules

File: app/javascript/src/styles/pages/_user.scss [hunk 7]
```diff
@@ -0,0 +1,30 @@
+.event-visible-item{
```
Comment (VasylYaremkiv) [inline app/javascript/src/styles/pages/_user.scss]: missed space

File: spec/controllers/targets_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+RSpec.describe TargetsController, type: :controller do
+  describe 'GET #change_achieved' do
+    let!(:user) { FactoryBot.create(:user) }
+    let!(:event) { FactoryBot.create(:event, user: user) }
+    let!(:target) { FactoryBot.create(:target, event: event) }
+
+    before(:each) do
+      login_user user
+    end
+    it 'change achieved field from false to true' do
```
Comment (VasylYaremkiv) [inline spec/controllers/targets_controller_spec.rb]: extra line

File: spec/controllers/targets_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+      expect(target.achieved).to be false
+
+      params = { id: target.id }
+      get :change_achieved, params: params, format: :json
+      target.reload
+
+      expect(target.achieved).to be true
+    end
+    it 'change achieved field from true to false' do
+      target.achieved = true
```
Comment (VasylYaremkiv) [inline spec/controllers/targets_controller_spec.rb]: `target.update_attributes(achieved: true)`

File: app/helpers/events_helper.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+module EventsHelper
+  def as_json_events_for_user(events)
```
Comment (VasylYaremkiv) [inline app/helpers/events_helper.rb]: Move it to model.

File: spec/features/user_profile_spec.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+  it 'show user data' do
+    expect(page).to have_content user.first_name
+    expect(page).to have_content user.last_name
+    expect(page).to have_content user.phone
+    expect(page).to have_content user.project.name
+    expect(page).to have_content user.project.desc
+  end
+
+  it 'show sorted event' do
```
Comment (VasylYaremkiv) [inline spec/features/user_profile_spec.rb]: `show sorted events`

File: spec/features/user_profile_spec.rb [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    finish_date_of_events_from_page = page.all('.event-finish-date').map(&:text)
```
Comment (VasylYaremkiv) [inline spec/features/user_profile_spec.rb]: Rename `finish_date_of_events_from_page` to `page_events` and move it after expected_finish_date_list

File: spec/features/user_profile_spec.rb [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    expected_finish_date_list = [first_event.finish_date.iso8601,
```
Comment (VasylYaremkiv) [inline spec/features/user_profile_spec.rb]: Rename to `expected_events`

File: spec/controllers/targets_controller_spec.rb (hunk 8 above)
Comment (VasylYaremkiv) [inline spec/controllers/targets_controller_spec.rb]: Rename to `mark as achieved if target is unachieved`

File: spec/controllers/targets_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      target.save
+      expect(target.achieved).to be true
+
+      params = { id: target.id }
+      get :change_achieved, params: params, format: :json
+      target.reload
+
+      expect(target.achieved).to be false
```
Comment (VasylYaremkiv) [inline spec/controllers/targets_controller_spec.rb]: Use `target.achieved?`

File: app/controllers/targets_controller.rb [hunk 15]
```diff
@@ -30,6 +30,17 @@ def destroy
     end
   end
 
+  def change_achieved
```
Comment (VasylYaremkiv) [inline app/controllers/targets_controller.rb]: Rename to `toggle_achieved`

File: spec/features/user_profile_spec.rb [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    expect(page).to have_content second_event.finish_date.year
+
+    expect(page).to_not have_content first_event.start_date
+    expect(page).to_not have_content first_event.comments
+    expect(page).to_not have_content first_event.summary
+
+    expect(expected_finish_date_list).to eq(finish_date_of_events_from_page)
+  end
+
+  it 'show hidden event data when click on event header' do
```
Comment (VasylYaremkiv) [inline spec/features/user_profile_spec.rb]: Did not get this test. You clicked on first event, but you check if first and second event info  present.

File: config/routes.rb [hunk 17]
```diff
@@ -4,5 +4,6 @@
                      path_names: { sign_in: 'login', sign_out: 'logout' }
   root to: 'users#index'
   resources :users, only: %i[edit update show index]
+  get 'change_achieved', action: :change_achieved, controller: 'targets'
```
Comment (VasylYaremkiv) [inline config/routes.rb]: You changed object.
If you want to retrieve object, use method `get`.
If you want to create object, use method `post`.
If you want to update object, use method `put`.

File: spec/features/user_profile_spec.rb [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    expect(page).to have_content first_event.start_date
+    expect(page).to have_content first_event.comments
+    expect(page).to have_content first_event.summary
+
+    expect(page).to have_content second_event.start_date
+    expect(page).to have_content second_event.comments
+    expect(page).to have_content second_event.summary
+  end
+
+  it 'change achieved status and save it' do
```
Comment (VasylYaremkiv) [inline spec/features/user_profile_spec.rb]: `mark target as achieved`

File: config/routes.rb [hunk 19]
```diff
@@ -4,5 +4,6 @@
                      path_names: { sign_in: 'login', sign_out: 'logout' }
   root to: 'users#index'
   resources :users, only: %i[edit update show index]
+  put 'target/:id/toggle_achieved', action: :toggle_achieved, controller: 'targets', as: 'toggle_achieved'
```
Comment (VasylYaremkiv) [inline config/routes.rb]:
```
resources :targets, only: [] do
  put 'toggle_achieved' , on: :member
end
```

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -27,6 +27,8 @@ def update
 
   def show
     @user = params[:id] ? User.find(params[:id]) : current_user
+    @user_project = Project.find(@user.project_id)
+    @events = Event.to_json(@user.events.order(finish_date: :desc))
```
Comment (VasylYaremkiv) [inline app/controllers/users_controller.rb]: @events = @user.events.sorted.map(&:to_json)

File: app/models/event.rb [hunk 21]
```diff
@@ -2,4 +2,21 @@ class Event < ApplicationRecord
   has_many :targets, dependent: :destroy
   belongs_to :user
   enum status: %i[scheduled done]
+
+  def self.to_json(events)
```
Comment (VasylYaremkiv) [inline app/models/event.rb]: def as_json
end

**Comment by SetterVet** [comment]
Ready to review @VasylYaremkiv @sonishc @rubycop @afurm

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -27,6 +27,8 @@ def update
 
   def show
     @user = params[:id] ? User.find(params[:id]) : current_user
+    @user_project = Project.find(@user.project_id)
+    @events = @user.events.order(finish_date: :desc).map(&:data)
```
Comment (VasylYaremkiv) [inline app/controllers/users_controller.rb]: move to scope `order(finish_date: :desc)`

File: app/views/users/show.html.haml [hunk 23]
```diff
@@ -1,8 +1,3 @@
-%br
-#user-profile
-  %h1 Profile page
-  %h5#profile-first_name= @user.first_name
-  %h5#profile-last_name= @user.last_name
-  %h5#profile-user_phone= @user.phone
-  %h5#profile-email= @user.email
-  %h5#profile-project= @user.project.name
+.users_component
+  %user-info-component{ "user-for-props": @user.to_json, "user-project-for-props":@user_project.to_json }
```
Comment (VasylYaremkiv) [inline app/views/users/show.html.haml]: `@user.project.to_json`

File: app/javascript/components/users/UserEvents.vue [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+        titlesEvent: {
+          "Comments": "comments",
+          "Summary": "summary",
+          "Start date": "start_date",
+          "Finish date": "finish_date",
+        },
+      }
+    },
+    props: {
+      eventsForProps: {type: String},
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/UserEvents.vue]: add space

File: app/models/user.rb [hunk 25]
```diff
@@ -14,23 +14,20 @@ def self.filter(id)
...
+  def data
+    user_project_info = project
+    {
+      id: id,
+      first_name: first_name,
+      last_name: last_name,
+      email: email,
+      phone: phone,
+      project: {
+        id: user_project_info.id,
```
Comment (VasylYaremkiv) [inline app/models/user.rb]: use `project`

File: app/views/users/show.html.haml (hunk 23 above)
Comment (VasylYaremkiv) [inline app/views/users/show.html.haml]: add space

File: config/routes.rb [hunk 26]
```diff
@@ -4,5 +4,10 @@
                      path_names: { sign_in: 'login', sign_out: 'logout' }
   root to: 'users#index'
   resources :users, only: %i[edit update show index]
+
   get 'profile', action: :show, controller: 'users'
+
+  resources :targets, only: %i[toggle_achieved] do
```
Comment (VasylYaremkiv) [inline config/routes.rb]: resources :targets, only: [] do

File: spec/controllers/targets_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    let!(:user) { FactoryBot.create(:user) }
+    let!(:event) { FactoryBot.create(:event, user: user) }
+    let!(:target) { FactoryBot.create(:target, event: event) }
+
+    before(:each) do
+      login_user user
+    end
+
+    it 'mark as achieved if target is unachieved' do
+      expect(!target.achieved?)
```
Comment (VasylYaremkiv) [inline spec/controllers/targets_controller_spec.rb]: expect(target.achieved?).to be_falsey

**Comment by SetterVet** [comment]
ready to Review @VasylYaremkiv @afurm @rubycop @sonishc

File: spec/models/user_spec.rb [hunk 28]
```diff
@@ -30,13 +30,13 @@
...
 
-  describe '.as_json' do
+  describe 'data' do
     context 'when array of instances converting to json' do
       let(:project) { create(:project) }
       let(:user) { create(:user, project: project) }
 
       it 'returns an array of objects as json' do
-        as_json = JSON.parse(User.as_json([user])).first
+        as_json = JSON.parse(user.data.to_json)
```
Comment (VasylYaremkiv) [inline spec/models/user_spec.rb]: You do not need to user `to_json` and after parse this json.
`user_data = user.data`

**Pull request merged**